ThingIFSDK/ThingIFSDK/Gateway/GatewayAPI.swift

```diff
@@ -0,0 +1,529 @@
+
+    private static let SHARED_NSUSERDEFAULT_KEY_INSTANCE = "GatewayAPI_INSTANCE"
+    private static func getSharedNSDefaultKey(tag : String?) -> String{
+        return SHARED_NSUSERDEFAULT_KEY_INSTANCE + (tag == nil ? "" : "_\(tag)")
+    }
+
+    public let tag: String?
+    public let app: App
+    public let gatewayAddress: NSURL
+
+    internal var accessToken: String?
+
+    let operationQueue = OperationQueue()
+
+    // MARK: - Implements NSCoding protocol
+    public func encodeWithCoder(aCoder: NSCoder)
+    {
+        aCoder.encodeObject(self.tag, forKey: "tag")
+        aCoder.encodeObject(self.app, forKey: "app")
+        aCoder.encodeObject(self.gatewayAddress, forKey: "gatewayAddress")
+        aCoder.encodeObject(self.accessToken, forKey: "accessToken")
+    }
+
+    public required init(coder aDecoder: NSCoder)
+    {
+        self.tag = aDecoder.decodeObjectForKey("tag") as? String
+        self.app = aDecoder.decodeObjectForKey("app") as! App
+        self.gatewayAddress = aDecoder.decodeObjectForKey("gatewayAddress") as! NSURL
+        self.accessToken = aDecoder.decodeObjectForKey("accessToken") as? String
+    }
+
+    init(app: App, gatewayAddress: NSURL, tag: String? = nil)
+    {
+        self.tag = tag
+        self.app = app
+        self.gatewayAddress = gatewayAddress
+    }
+
+    // MARK: API methods
+
+    /** Login to the Gateway.
+     Local authentication for the Gateway access.
+     Required prior to call other APIs access to the gateway.
+
+     - Parameter username: Username of the Gateway.
+     - Parameter password: Password of the Gateway.
+     - Parameter completionHandler: A closure to be executed once finished. The closure takes 1 argument: an instance of ThingIFError when failed.
+     */
+    public func login(
+        username: String,
+        password: String,
+        completionHandler: (ThingIFError?)-> Void
+        )
+    {
+        if username.isEmpty || password.isEmpty {
+            completionHandler(ThingIFError.UNSUPPORTED_ERROR)
+            return
+        }
+
+        let requestURL = "\(self.app.baseURL)/\(self.app.siteName)/token"
+
+        // generate header
+        let credential = "\(self.app.appID):\(self.app.appKey)"
+        let base64Str = credential.dataUsingEncoding(NSUTF8StringEncoding)?.base64EncodedDataWithOptions(NSDataBase64EncodingOptions.Encoding64CharacterLineLength)
+        let requestHeaderDict:Dictionary<String, String> = [
+            "authorization": "Bearer \(base64Str)"
+        ]
+
+        // genrate body
+        let requestBodyDict = NSMutableDictionary(dictionary:
+            [
+                "username": username,
+                "password": password
+            ]
+        )
+
+        do {
+            let requestBodyData = try NSJSONSerialization.dataWithJSONObject(requestBodyDict, options: NSJSONWritingOptions(rawValue: 0))
+            // do request
+            let request = buildDefaultRequest(
```

Comment:
not to use `buildDefaultRequest` to access gateway api, since the `X-Kii-SDK` header is not needed. Please create new builder for it.